website/blog/2020-07-17-unnest.md

```diff
@@ -0,0 +1,219 @@
+
+## Opportunities for Improvements? Yes!
+
+The baseline implementation did a number of things well. It avoided deep copy by using DictionaryBlock which improved the original improvement by an order of magnitude for cases without NULL insertions. Can we improve it additionally? The answer is yes.
+
+### Process data column-by-column, not row-by-row
+If you read our blog post https://prestodb.io/blog/2019/12/20/repartition, you may remember the 5 design choices we picked for repartitioning. The first was “Process data column by column, not row by row”. The same thing applies to `UnnestOperator` too. The baseline implementation constructs the output blocks in the row-by-row manner:
+
+```text
+for each incoming row
+    for each replicated column
+        append the repeated values by adding ids to the DictionaryBlock
+    for each unnest column
+        append the unnested  values by adding ids to the DictionaryBlock if possible.
+        Otherwise append the value to BlockBuilder.
+    if there is ordinality column
+        append the oridinality to the BlockBuilder
+```
+
+As a key principle of vectorized execution, we want to optimize the operations in a column by column manner, which uses tight loops and allows the JVM to vectorize the binaries and the CPU to parallelize the work as much as possible. An example of such loop of building the replicated block is like this:
+
+```java
+public Block buildOutputBlock(int[] maxEntries, int offset, int length, int totalEntries)
+{
+    int[] ids = new int[totalEntries];
+    int fromPosition = 0;
+    for (int i = 0; i < length; i++) {
+        int toPosition = fromPosition + maxEntries[offset + i];
+        Arrays.fill(ids, fromPosition, toPosition, sourcePosition++);
+        fromPosition = toPosition;
+    }
+
+    return new DictionaryBlock(totalEntries, source, ids);
+}
+```
+
+This simple change itself improved the performance for the NO-NULL case for at least 3x as shown in the JMH benchmarks.
+
+### Computation of maxEntries array and batch size
+
+To make the above loop possible, we pre-compute the cardinalities for each row for all unnest columns and use them to get the max cardinalities (we call it the maxEntries array) in the operator. This way the memory can be accessed sequentially and the computation can be reused. The size of these arrays should require very little memory overhead, since they’re only for top level rows and should always be less than 1024 per block.
+
+We then compute the batch size, ie. top level row count that can fit into the next output blocks. Right now we limit the nested row count for each batch 1000. The batch size is at least 1. This means the unnested row count for each block might be over 1000. For example, if a row contains 10000 array entries, then the unnested row count would be 10000. Although this is over the 1000 rows we still have to output the whole top level row which translates into 10000 unnested rows.
```

Comment:
output blocks or output block? It's outputting one block at a time so probably singular?

"Right now we limit the nested row count for each batch to 1000 with a minimum batch size of 1" - added "to 1000" and combined the two sentences.

Reply:
> output blocks or output block? It's outputting one block at a time so probably singular?
> 
It's blocks. We calculate batch size for all blocks. To avoid confusion, I changed it to `output page`.